contracts/sponsorship/BiconomySponsorshipPaymaster.sol

```diff
@@ -261,25 +265,31 @@ contract BiconomySponsorshipPaymaster is
         override
     {
         unchecked {
-            (address paymasterId, uint32 priceMarkup, bytes32 userOpHash, uint256 prechargedAmount) =
-                abi.decode(context, (address, uint32, bytes32, uint256));
+            (address paymasterId, uint32 priceMarkup, uint256 prechargedAmount) =
+                abi.decode(context, (address, uint32, uint256));
 
             // Include unaccountedGas since EP doesn't include this in actualGasCost
             // unaccountedGas = postOpGas + EP overhead gas + estimated penalty
             actualGasCost = actualGasCost + (unaccountedGas * actualUserOpFeePerGas);
             // Apply the price markup
             uint256 adjustedGasCost = (actualGasCost * priceMarkup) / _PRICE_DENOMINATOR;
 
+            uint256 premium = adjustedGasCost - actualGasCost;
+
+            // Add priceMarkup to fee collector balance
+            paymasterIdBalances[feeCollector] += premium;
+
             if (prechargedAmount > adjustedGasCost) {
                 // If overcharged refund the excess
                 paymasterIdBalances[paymasterId] += (prechargedAmount - adjustedGasCost);
+                // here adjustedGasCost does not account for gasPenalty
+                emit GasBalanceDeducted(paymasterId, adjustedGasCost, premium);
+            } else {
+                // deduct what needs to be deducted from paymasterId
+                paymasterIdBalances[paymasterId] -= (adjustedGasCost - prechargedAmount);
+                // here chargedAmount accounts for penalty with maxGasPenalty
+                emit GasBalanceDeducted(paymasterId, prechargedAmount, premium);
```

Comment:
I believe you should emit `adjustedGasCost` in this case as it is now what in fact was charged after L289 right?
so `emit` can be moved out of if/else

Reply:
right. with this addition it makes sense
I wanted to emit different things that's the whole reason I added if else. 
but it's ok. 
knowing that charge can never be accurately emitted here and will need some off-chain component for penalties.